rust/feature-flags/src/flag_matching.rs

```diff
@@ -228,101 +546,245 @@ impl FeatureFlagMatcher {
-        let hash = self.get_hash(feature_flag, "variant");
-        let mut total_percentage = 0.0;
-
-        for variant in feature_flag.get_variants() {
-            total_percentage += variant.rollout_percentage / 100.0;
-            if hash < total_percentage {
-                return Some(variant.key.clone());
-            }
+    async fn check_rollout(
+        &mut self,
+        feature_flag: &FeatureFlag,
+        rollout_percentage: f64,
+    ) -> Result<(bool, String), FlagError> {
+        if rollout_percentage == 100.0
+            || self.get_hash(feature_flag, "").await? <= (rollout_percentage / 100.0)
+        {
+            Ok((true, "CONDITION_MATCH".to_string())) // TODO enum, I'll implement this when I implement evaluation reasons
+        } else {
+            Ok((false, "OUT_OF_ROLLOUT_BOUND".to_string())) // TODO enum, I'll implement this when I implement evaluation reasons
         }
-        None
     }
 
     /// This function takes a feature flag and returns the key of the variant that should be shown to the user.
-    pub async fn get_person_properties_from_cache_or_db(
+    async fn get_matching_variant(
         &mut self,
-        team_id: i32,
-        distinct_id: String,
-    ) -> Result<HashMap<String, Value>, FlagError> {
-        // TODO: Do we even need to cache here anymore?
-        // Depends on how often we're calling this function
-        // to match all flags for a single person
-
-        // TODO which of these properties do we need to cache?
-        if let Some(cached_props) = self.cached_properties.clone() {
-            // TODO: Maybe we don't want to copy around all user properties, this will by far be the largest chunk
-            // of data we're copying around. Can we work with references here?
-            // Worst case, just use a Rc.
-            return Ok(cached_props);
-        }
+        feature_flag: &FeatureFlag,
+    ) -> Result<Option<String>, FlagError> {
+        let hash = self.get_hash(feature_flag, "variant").await?;
+        let mut cumulative_percentage = 0.0;
 
-        if self.database_client.is_none() {
-            return Err(FlagError::DatabaseUnavailable);
+        for variant in feature_flag.get_variants() {
+            cumulative_percentage += variant.rollout_percentage / 100.0;
+            if hash < cumulative_percentage {
+                return Ok(Some(variant.key.clone()));
+            }
         }
+        Ok(None)
+    }
+}
 
-        let mut conn = self
-            .database_client
-            .as_ref()
-            .expect("client should exist here")
-            .get_connection()
-            .await?;
+async fn fetch_and_locally_cache_all_properties(
+    properties_cache: &mut PropertiesCache,
+    database_client: Option<DatabaseClientArc>,
+    distinct_id: String,
+    team_id: TeamId,
+    group_type_indexes: &HashSet<GroupTypeIndex>,
+) -> Result<(), FlagError> {
+    if database_client.is_none() {
```

Comment:
You can use `let Some(db_client) = database_client else { return Err }` here to avoid the unwraps you do below, but I kind of wonder why you let an `Option` be passed here at all if it's always invalid to do so (this is done in a bunch of places all over here, with mandatory arguments for the function not to return an error being passed as `Option`s, and I think it muddles your code a lot)